README.md
# chilexpress

[![npm version](https://img.shields.io/npm/v/chilexpress.svg?style=flat-square)](https://www.npmjs.com/package/chilexpress)
[![npm downloads](https://img.shields.io/npm/dm/chilexpress.svg?style=flat-square)](https://www.npmjs.com/package/chilexpress)
[![Build Status](https://travis-ci.org/lgaticaq/chilexpress.svg?branch=master)](https://travis-ci.org/lgaticaq/chilexpress)
[![Maintainability](https://api.codeclimate.com/v1/badges/d8aabe323571c8fb5f0e/maintainability)](https://codeclimate.com/github/lgaticaq/chilexpress/maintainability)
[![dependency Status](https://img.shields.io/david/lgaticaq/chilexpress.svg?style=flat-square)](https://david-dm.org/lgaticaq/chilexpress#info=dependencies)
[![devDependency Status](https://img.shields.io/david/dev/lgaticaq/chilexpress.svg?style=flat-square)](https://david-dm.org/lgaticaq/chilexpress#info=devDependencies)

> Check shipping status in chilexpress

## Installation

```bash
npm i -S chilexpress
```

## Use

[Try on Tonic](https://tonicdev.com/npm/chilexpress)
```js
import chilexpress from 'chilexpress';

const orderId = '111111111111';

chilexpress(orderId).then(console.log);
/*
{
  status: 'En Proceso De Despacho',
  delivery: {
    name: 'XXX XXX',
    rut: '0',
    date: Mon Dec 28 2015 11:31:00 GMT+0000 (UTC),
    signature: ''
  },
  history: [
    {
      date: Mon Dec 28 2015 11:31:00 GMT+0000 (UTC),
      activity: 'En Proceso De Despacho'
    }
  ],
  shipping: {
    product: 'Encomienda',
    service: 'Dia Habil Siguiente',
    date: Mon Dec 28 2015 11:31:00 GMT+0000 (UTC)
  }
}
*/
```

## License

[MIT](https://tldrlegal.com/license/mit-license)


CHANGELOG.md
## [3.0.4](https://github.com/lgaticaq/chilexpress/compare/v3.0.3...v3.0.4) (2021-04-19)


### Bug Fixes

* **deps:** update dependency chrome-aws-lambda to ^3.1.1 ([e5a5f8a](https://github.com/lgaticaq/chilexpress/commit/e5a5f8a389b374ab2c1f4d5665eb65db83c52235))

## [3.0.3](https://github.com/lgaticaq/chilexpress/compare/v3.0.2...v3.0.3) (2020-11-19)


### Bug Fixes

* **deps:** update dependency debug to v4.3.0 ([9b03414](https://github.com/lgaticaq/chilexpress/commit/9b03414dbc3a73480a6d58caf4e7f072e7c6f972))
* **deps:** update dependency debug to v4.3.1 ([c7b5377](https://github.com/lgaticaq/chilexpress/commit/c7b537783f9b1989b85ddb3f680802ed76c0056d))

## [3.0.2](https://github.com/lgaticaq/chilexpress/compare/v3.0.1...v3.0.2) (2020-09-19)


### Bug Fixes

* **deps:** update dependency debug to v4.2.0 ([a184703](https://github.com/lgaticaq/chilexpress/commit/a18470360f51a1fea72d9ace0956ff964c9bb334))

## [3.0.1](https://github.com/lgaticaq/chilexpress/compare/v3.0.0...v3.0.1) (2020-06-02)


### Bug Fixes

* **deps:** update dependency chrome-aws-lambda to v3 ([19dc028](https://github.com/lgaticaq/chilexpress/commit/19dc028ad203a3029e23dbe9871a67d388e1ade5))

# [3.0.0](https://github.com/lgaticaq/chilexpress/compare/v2.0.6...v3.0.0) (2019-11-20)


### Bug Fixes

* **linter:** Fix Arrow function should not return assignment ([a73d689](https://github.com/lgaticaq/chilexpress/commit/a73d689fc50e0003d9c31ab2d3045a36294ef9a6))
* **package:** update cheerio to version 1.0.0-rc.1 ([574c6d4](https://github.com/lgaticaq/chilexpress/commit/574c6d4537807f0fa4294a14c3d8727c0e916500))


### Features

* migrate to puppeteer ([2709bd7](https://github.com/lgaticaq/chilexpress/commit/2709bd7d388385e8398beb93279979ebb542ef37))


### BREAKING CHANGES

* new output data

Before:
```js
chilexpress('111111111111').then(console.log)
/*
{
  orderId: '111111111111',
  product: 'Encomienda',
  service: 'Dia Habil Siguiente',
  status: 'En Proceso De Despacho',
  isDeliveried: false,
  history: [
    {
      datetime: Mon Dec 28 2015 11:31:00 GMT+0000 (UTC),
      activity: 'En Proceso De Despacho'
    }
  ]
}
*/
```

After:
```js
chilexpress('111111111111').then(console.log);
/*
{
  status: 'En Proceso De Despacho',
  delivery: {
    name: 'XXX XXX',
    rut: '0',
    date: Mon Dec 28 2015 11:31:00 GMT+0000 (UTC),
    signature: ''
  },
  history: [
    {
      date: Mon Dec 28 2015 11:31:00 GMT+0000 (UTC),
      activity: 'En Proceso De Despacho'
    }
  ],
  shipping: {
    product: 'Encomienda',
    service: 'Dia Habil Siguiente',
    date: Mon Dec 28 2015 11:31:00 GMT+0000 (UTC)
  }
}
*/
```

#### 2.0.6 (2017-3-14)

##### Chores

* **package:**
  * Upgrade eslint to v3.17.1 ([2762ab69](https://github.com/lgaticaq/chilexpress/commit/2762ab699d2d3182fe5460e3166f269402ecfde9))
  * Upgrade coveralls to v2.12.0 ([302aa916](https://github.com/lgaticaq/chilexpress/commit/302aa9164ba0b563a776f6caaabc721aadd6ef9f))
  * Upgrade codeclimate-test-reporter to v0.4.1 ([036b2655](https://github.com/lgaticaq/chilexpress/commit/036b26550b6005370bb3c068e706477d7ceb78fa))
  * update nock to version 9.0.9 ([dc11f6f7](https://github.com/lgaticaq/chilexpress/commit/dc11f6f7815dfd2a55fbfd83e3593efcf17bbb16))
  * Update eslint to v3.14.1 ([8ad8317f](https://github.com/lgaticaq/chilexpress/commit/8ad8317fe57db10ccfa42547ba2b29d9a1e93278))
* **yarn:** Update lock ([8a75df8d](https://github.com/lgaticaq/chilexpress/commit/8a75df8d4516520dcf553377a6ca95cdcc6f5dda))
* **travis:** Add cache yarn ([b9befe55](https://github.com/lgaticaq/chilexpress/commit/b9befe55562ad5f3b9944276d78817072633e1a0))

#### 2.0.5 (2016-12-18)

##### Bug Fixes

* **datetime:** Fix timezone ([a7256ee5](https://github.com/lgaticaq/chilexpress/commit/a7256ee5f9d40781e8d383e0a94b4fe1c17aa167))

#### 2.0.4 (2016-12-18)

##### Chores

* **travis:** Update node version ([d1c1b4c5](https://github.com/lgaticaq/chilexpress/commit/d1c1b4c5ffe67174ec1620433ef7fbeded86805c))
* **nvm:** Update node version ([2c1c6325](https://github.com/lgaticaq/chilexpress/commit/2c1c6325eb0b6465640c9cff706e12b9dbc3d6a4))
* **npm:** Update ignore ([ef13b6b7](https://github.com/lgaticaq/chilexpress/commit/ef13b6b7091665bd3e6b2d07a9b7077201cc5854))
* **package:**
  * Add yarn.lock ([5b04a369](https://github.com/lgaticaq/chilexpress/commit/5b04a36924681cbde8722d150dd022942adcc8b1))
  * Update dependencies ([4bfcbd36](https://github.com/lgaticaq/chilexpress/commit/4bfcbd366c955f207957e4ce1f3accb421646bef))
  * Replace request-promise with https ([54f43946](https://github.com/lgaticaq/chilexpress/commit/54f43946e36e72b0c75ef9d6d9d7ae0f332bc7cc))
  * update nock to version 9.0.0 ([9717e425](https://github.com/lgaticaq/chilexpress/commit/9717e42594ec28460499f5287051fe7b8fb4a5b5))
  * update codeclimate-test-reporter to version 0.4.0 ([962cb33c](https://github.com/lgaticaq/chilexpress/commit/962cb33c1318a4e3435a7865419678f330373965))
  * update cheerio to version 0.22.0 ([d8c4c911](https://github.com/lgaticaq/chilexpress/commit/d8c4c9110d25980e005061dcb205cc7d8a2e3465))
* **tests:** Complete tests ([e979d2a0](https://github.com/lgaticaq/chilexpress/commit/e979d2a09ae952511a42f83d1ee9c4e3875f4269))

#### 2.0.3 (2016-8-2)

##### Chores

* **package:**
  * Update dependencies ([626e3ee3](https://github.com/lgaticaq/chilexpress/commit/626e3ee3f5763c4d803504040ad6f0e260d199b3))
  * update mocha to version 3.0.0 ([7c382ab9](https://github.com/lgaticaq/chilexpress/commit/7c382ab9d1bf47d0affc0fd4988a18c252a47c5b))
  * update request-promise to version 4.0.1 ([dd592194](https://github.com/lgaticaq/chilexpress/commit/dd5921944da7ebd9dc1acdd35d73de69b63b2ea3))
  * update eslint to version 3.0.0 ([3b3a6e11](https://github.com/lgaticaq/chilexpress/commit/3b3a6e111892ef8d50cfd014923d0bf88723a7f3))
  * update eslint to version 2.13.0 ([f95b3fd6](https://github.com/lgaticaq/chilexpress/commit/f95b3fd661cdd1e6093d1cb2f16cb6297dfa21f2))

test/test.js
'use strict'

const path = require('path')

const { expect } = require('chai')
const nock = require('nock')

const lib = require('../src')

describe('chilexpress', () => {
  describe('valid order id', () => {
    beforeEach(() => {
      nock.disableNetConnect()
      nock('https://www.chilexpress.cl')
        .get('/Views/ChilexpressCL/Resultado-busqueda.aspx')
        .query({ DATA: '111111111111' })
        .replyWithFile(200, path.join(__dirname, 'valid.html'))
    })

    it('should return a array statuses', done => {
      lib('111111111111')
        .then(data => {
          expect(data).to.be.a('object')
          expect(data).to.include.keys(
            'orderId',
            'product',
            'service',
            'status',
            'isDeliveried',
            'history'
          )
          for (let history of data.history) {
            expect(history).to.include.keys('datetime', 'activity')
            expect(history.datetime).to.be.a('date')
            expect(history.activity).to.be.a('string')
          }
          done()
        })
        .catch(err => {
          expect(err).to.eql(null)
          done()
        })
    })
  })

  describe('valid order id with delivery', () => {
    beforeEach(() => {
      nock.disableNetConnect()
      nock('https://www.chilexpress.cl')
        .get('/Views/ChilexpressCL/Resultado-busqueda.aspx')
        .query({ DATA: '111111111111' })
        .replyWithFile(200, path.join(__dirname, 'delivery.html'))
    })

    it('should return a array statuses', done => {
      lib('111111111111')
        .then(data => {
          expect(data).to.be.a('object')
          expect(data).to.include.keys(
            'orderId',
            'product',
            'service',
            'status',
            'isDeliveried',
            'history'
          )
          for (let history of data.history) {
            expect(history).to.include.keys('datetime', 'activity')
            expect(history.datetime).to.be.a('date')
            expect(history.activity).to.be.a('string')
          }
          done()
        })
        .catch(err => {
          expect(err).to.eql(null)
          done()
        })
    })
  })

  describe('invalid order id', () => {
    beforeEach(() => {
      nock.disableNetConnect()
      nock('https://www.chilexpress.cl')
        .get('/Views/ChilexpressCL/Resultado-busqueda.aspx')
        .query({ DATA: '1' })
        .replyWithFile(200, path.join(__dirname, 'invalid.html'))
    })

    it('should return an error', done => {
      lib('1').catch(err => {
        expect(err).to.be.an('error')
        done()
      })
    })
  })

  describe('server error', () => {
    beforeEach(() => {
      nock.disableNetConnect()
      nock('https://www.chilexpress.cl')
        .get('/Views/ChilexpressCL/Resultado-busqueda.aspx')
        .query({ DATA: '1' })
        .replyWithError('Server error')
    })

    it('should return an error', done => {
      lib('1').catch(err => {
        expect(err).to.be.an('error')
        done()
      })
    })
  })

  describe('bad status code', () => {
    beforeEach(() => {
      nock.disableNetConnect()
      nock('https://www.chilexpress.cl')
        .get('/Views/ChilexpressCL/Resultado-busqueda.aspx')
        .query({ DATA: '1' })
        .reply(301)
    })

    it('should return an error', done => {
      lib('1').catch(err => {
        expect(err).to.be.an('error')
        done()
      })
    })
  })
})


src/index.js
const puppeteer = require('puppeteer')

/**
 * @typedef {Object} ChilexpressDelivery
 * @property {string} name
 * @property {string} rut
 * @property {string} date
 * @property {string} signature
 */
/**
 * @typedef {Object} ChilexpressHistory
 * @property {string} date
 * @property {string} activity
 */
/**
 * @typedef {Object} ChilexpressShipping
 * @property {string} product
 * @property {string} service
 * @property {string} date
 */
/**
 * @typedef {Object} ChilexpressResult
 * @property {string} status
 * @property {ChilexpressDelivery} delivery
 * @property {Array<ChilexpressHistory>} history
 * @property {ChilexpressShipping} shipping
 */
/**
 * @param {number} orderId - Chilexpress tracking id.
 * @returns {ChilexpressResult} -
 * @example
 * const result = await chilexpress(orderId)
 */
module.exports = async orderId => {
  const browser = await puppeteer.launch()
  const page = await browser.newPage()

  const navigationPromise = page.waitForNavigation()

  await page.goto(
    'https://www.chilexpress.cl/Views/ChilexpressCL/Resultado-busqueda.aspx'
  )

  await page.setViewport({ width: 1042, height: 630 })

  await navigationPromise

  // Search order
  await page.waitForSelector('#Cabecera_divContenedorMenu #Cabecera_txtSearch')
  await page.type('#Cabecera_divContenedorMenu #Cabecera_txtSearch', orderId)
  await page.waitForSelector(
    '.row > .grid-13 > .form-inline > #Cabecera_btnBuscar > .icon-search'
  )
  await page.click(
    '.row > .grid-13 > .form-inline > #Cabecera_btnBuscar > .icon-search'
  )

  // Get status content
  await page.waitForSelector(
    '.seguimiento_grupo > .seguimiento_imagen.seguimiento_verde.seguimiento_paso_actual'
  )
  const status = await page.evaluate(
    () =>
      document
        .querySelector(
          '.seguimiento_grupo > .seguimiento_imagen.seguimiento_verde.seguimiento_paso_actual'
        )
        .parentElement.getElementsByTagName('div')[0].textContent
  )

  // Get delivery content
  await page.waitForSelector('.seccion > .datos_entrega > .contenido_entrega')
  /** @type {ChilexpressDelivery} */
  const delivery = await page.evaluate(() => ({
    name: document
      .querySelector(
        '.seccion > .datos_entrega > .contenido_entrega > .grupo_datos:nth-child(1) > .datos_informacion'
      )
      .textContent.trim(),
    rut: document
      .querySelector(
        '.seccion > .datos_entrega > .contenido_entrega > .grupo_datos:nth-child(2) > .datos_informacion'
      )
      .textContent.trim(),
    date: document
      .querySelector(
        '.seccion > .datos_entrega > .contenido_entrega > .grupo_datos:nth-child(3) > .datos_informacion'
      )
      .textContent.trim()
      .replace(
        /(\d{2})\/(\d{2})\/(\d{4}) - (\d{2}):(\d{2})/,
        '$3-$2-$1T$4:$5:00-03:00'
      ),
    signature: document
      .querySelector(
        '.seccion > .datos_entrega > .contenido_entrega > .grupo_datos:nth-child(4) > .datos_informacion'
      )
      .textContent.trim()
  }))

  // Get history content
  await page.waitForSelector(
    '.datos_seguimiento > .addresses > #ListaTrackingOT'
  )
  /** @type {ChilexpressHistory} */
  const history = await page.evaluate(() => {
    return Array.from(
      document.querySelectorAll(
        '.datos_seguimiento > .addresses > #ListaTrackingOT > tr'
      )
    ).map(tr => {
      const [date, time, activity] = Array.from(tr.querySelectorAll('td')).map(
        td => td.textContent
      )
      return {
        date: `${date}${time}`.replace(
          /(\d{2})\/(\d{2})\/(\d{4})(\d{2}):(\d{2})/,
          '$3-$2-$1T$4:$5:00-03:00'
        ),
        activity
      }
    })
  })

  // Get shipping content
  await page.waitForSelector('.datos_entrega > .datos_pieza > .contenido_pieza')
  /** @type {ChilexpressShipping} */
  const shipping = await page.evaluate(() => ({
    product: document
      .querySelector(
        '.datos_entrega > .datos_pieza > .contenido_pieza > .grupo_datos:nth-child(1) > .pieza_producto'
      )
      .textContent.trim(),
    service: document
      .querySelector(
        '.datos_entrega > .datos_pieza > .contenido_pieza > .grupo_datos:nth-child(2) > .pieza_producto'
      )
      .textContent.trim(),
    date: document
      .querySelector(
        '.datos_entrega > .datos_pieza > .contenido_pieza > .grupo_datos:nth-child(3) > .pieza_producto'
      )
      .textContent.trim()
      .replace(/(\d{2})\/(\d{2})\/(\d{4})/, '$3-$2-$1T00:00:00-03:00')
  }))

  await browser.close()

  return { status, delivery, history, shipping }
}


src/chilexpress.js
const debug = require('debug')('chilexpress')

/**
 * @typedef {import('puppeteer').Browser} Browser
 */
/**
 * @typedef {Object} ChilexpressDelivery
 * @property {string} name
 * @property {string} rut
 * @property {Date} date
 * @property {string} signature
 */
/**
 * @typedef {Object} ChilexpressHistory
 * @property {Date} date
 * @property {string} activity
 */
/**
 * @typedef {Object} ChilexpressShipping
 * @property {string} product
 * @property {string} service
 * @property {Date} date
 */
/**
 * @typedef {Object} ChilexpressResult
 * @property {string} status
 * @property {ChilexpressDelivery} delivery
 * @property {Array<ChilexpressHistory>} history
 * @property {ChilexpressShipping} shipping
 */
/**
 * @param {Browser} browser - Puppeter browser instance.
 * @param {number} orderId - Chilexpress tracking id.
 * @returns {Promise<ChilexpressResult>} -
 * @example
 * const result = await chilexpress(orderId)
 */
module.exports = async (browser, orderId) => {
  debug('open new page')
  const page = await browser.newPage()

  const navigationPromise = page.waitForNavigation()

  debug(
    'got to https://www.chilexpress.cl/Views/ChilexpressCL/Resultado-busqueda.aspx'
  )
  await page.goto(
    'https://www.chilexpress.cl/Views/ChilexpressCL/Resultado-busqueda.aspx'
  )

  await page.setViewport({ width: 1042, height: 630 })

  debug('wait navigation')
  await navigationPromise

  debug('wait input search')
  await page.waitForSelector('#Cabecera_divContenedorMenu #Cabecera_txtSearch')
  debug(`type order id ${orderId}`)
  await page.type(
    '#Cabecera_divContenedorMenu #Cabecera_txtSearch',
    orderId.toString()
  )
  debug('wait button search')
  await page.waitForSelector(
    '.row > .grid-13 > .form-inline > #Cabecera_btnBuscar > .icon-search'
  )
  debug('click button search')
  await page.click(
    '.row > .grid-13 > .form-inline > #Cabecera_btnBuscar > .icon-search'
  )

  debug('wait content')
  await page.waitForSelector('#busqueda-content')
  const exist = await page.evaluate(() => {
    const header = document.querySelector('#busqueda-content > h2')
    return !header
  })
  debug(`content founded: ${exist}`)
  if (!exist) throw new Error('Order not found')

  debug('wait status content')
  await page.waitForSelector(
    '.seguimiento_grupo > .seguimiento_imagen.seguimiento_verde.seguimiento_paso_actual'
  )
  const status = await page.evaluate(
    () =>
      document
        .querySelector(
          '.seguimiento_grupo > .seguimiento_imagen.seguimiento_verde.seguimiento_paso_actual'
        )
        .parentElement.getElementsByTagName('div')[0].textContent
  )
  debug(`status content: ${status}`)

  debug('wait delivery content')
  await page.waitForSelector('.seccion > .datos_entrega > .contenido_entrega')
  /** @type {ChilexpressDelivery} */
  const delivery = await page.evaluate(() => ({
    name: document
      .querySelector(
        '.seccion > .datos_entrega > .contenido_entrega > .grupo_datos:nth-child(1) > .datos_informacion'
      )
      .textContent.trim(),
    rut: document
      .querySelector(
        '.seccion > .datos_entrega > .contenido_entrega > .grupo_datos:nth-child(2) > .datos_informacion'
      )
      .textContent.trim(),
    date: document
      .querySelector(
        '.seccion > .datos_entrega > .contenido_entrega > .grupo_datos:nth-child(3) > .datos_informacion'
      )
      .textContent.trim()
      .replace(
        /(\d{2})\/(\d{2})\/(\d{4}) - (\d{2}):(\d{2}) hrs\./,
        '$3-$2-$1T$4:$5:00-03:00'
      ),
    signature: document
      .querySelector(
        '.seccion > .datos_entrega > .contenido_entrega > .grupo_datos:nth-child(4) > .datos_informacion'
      )
      .textContent.trim()
  }))
  debug(`delivery content: ${JSON.stringify(delivery)}`)

  debug('wait history content')
  await page.waitForSelector(
    '.datos_seguimiento > .addresses > #ListaTrackingOT'
  )
  /** @type {Array<ChilexpressHistory>} */
  const history = await page.evaluate(() => {
    return Array.from(
      document.querySelectorAll(
        '.datos_seguimiento > .addresses > #ListaTrackingOT > tr'
      ),
      tr => {
        const [date, time, activity] = Array.from(
          tr.querySelectorAll('td'),
          td => td.textContent
        )
        return {
          date: `${date}${time}`.replace(
            /(\d{2})\/(\d{2})\/(\d{4})(\d{2}):(\d{2})/,
            '$3-$2-$1T$4:$5:00-03:00'
          ),
          activity
        }
      }
    )
  })
  debug(`history content: ${JSON.stringify(history)}`)

  debug('wait shipping content')
  await page.waitForSelector('.datos_entrega > .datos_pieza > .contenido_pieza')
  /** @type {ChilexpressShipping} */
  const shipping = await page.evaluate(() => ({
    product: document
      .querySelector(
        '.datos_entrega > .datos_pieza > .contenido_pieza > .grupo_datos:nth-child(1) > .pieza_producto'
      )
      .textContent.trim(),
    service: document
      .querySelector(
        '.datos_entrega > .datos_pieza > .contenido_pieza > .grupo_datos:nth-child(2) > .pieza_producto'
      )
      .textContent.trim(),
    date: document
      .querySelector(
        '.datos_entrega > .datos_pieza > .contenido_pieza > .grupo_datos:nth-child(3) > .pieza_producto'
      )
      .textContent.trim()
      .replace(/(\d{2})\/(\d{2})\/(\d{4})/, '$3-$2-$1T00:00:00-03:00')
  }))
  debug(`shipping content: ${JSON.stringify(shipping)}`)

  delivery.date = new Date(delivery.date)
  history.forEach(data => {
    data.date = new Date(data.date)
  })
  shipping.date = new Date(shipping.date)

  return { status, delivery, history, shipping }
}
